src/main/kotlin/com/pubnub/api/PubNub.kt

```diff
@@ -2105,4 +2239,211 @@ class PubNub internal constructor(
+     *
+     * This is the same as calling [PubNub.channel] followed by [Channel.subscription] for each channel,
+     * then creating a [subscriptionSetOf] using these [Subscription] objects (similarly for channel groups).
+     *
+     * @param channels the channels to create subscriptions for
+     * @param channelGroups the channel groups to create subscriptions for
+     * @param options the [SubscriptionOptions] to pass for each subscription. Refer to supported options in [Channel] and
+     * [ChannelGroup] documentation.
+     * @return a [SubscriptionSet] containing subscriptions for the given [channels] and [channelGroups]
+     */
+    fun subscriptionSetOf(
+        channels: Set<String> = emptySet(),
+        channelGroups: Set<String> = emptySet(),
+        options: SubscriptionOptions = SubscriptionOptions.Default
+    ): SubscriptionSet {
+        val subscriptionSet = SubscriptionSetImpl(this)
+        channels.forEach {
+            subscriptionSet.add(channel(it).subscription(options))
+        }
+        channelGroups.forEach {
+            subscriptionSet.add(channelGroup(it).subscription(options))
+        }
+        return subscriptionSet
+    }
+
+    // internal
+    private val LOCK_CHANNELS_AND_GROUPS = Any()
+    private val channelSubscriptions = mutableMapOf<ChannelName, MutableSet<Subscription>>()
+    private val channelGroupSubscriptions = mutableMapOf<ChannelGroupName, MutableSet<Subscription>>()
+
+    internal fun subscribe(vararg subscriptions: SubscriptionImpl, cursor: SubscriptionCursor) {
+        synchronized(LOCK_CHANNELS_AND_GROUPS) {
+            val channelsToSubscribe = mutableSetOf<ChannelName>()
+            subscriptions.forEach { subscription ->
+                subscription.channels.forEach { channelName ->
+                    channelSubscriptions.computeIfAbsent(channelName) { mutableSetOf() }
+                        .also { set -> set.add(subscription) }
+                    channelsToSubscribe.add(channelName)
+                }
+            }
+            val groupsToSubscribe = mutableSetOf<ChannelGroupName>()
+            subscriptions.forEach { subscription ->
+                subscription.channelGroups.forEach { channelGroupName ->
+                    channelGroupSubscriptions.computeIfAbsent(channelGroupName) { mutableSetOf() }
+                        .also { set -> set.add(subscription) }
+                    groupsToSubscribe.add(channelGroupName)
+                }
+            }
+
+            val (channelsWithPresence, channelsNoPresence) = channelsToSubscribe.filter { !it.isPresence }.partition {
+                channelsToSubscribe.contains(it.withPresence)
+            }
+            val (groupsWithPresence, groupsNoPresence) = groupsToSubscribe.filter { !it.isPresence }.partition {
+                groupsToSubscribe.contains(it.withPresence)
+            }
+            if (channelsWithPresence.isNotEmpty() || groupsWithPresence.isNotEmpty()) {
```

Comment:
As far as I understand there is possibility that we call `subscribeInternal` method twice.
Why we need to call `subscribeInternal` method twice?
Wouldn't it be more efficient to aggregate data and make just one call?